In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_df = pd.read_csv(
    Path('Resources/lending_data.csv')
)

# Review the DataFrame
lending_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features
# Separate the y variable, the labels
y = lending_df['loan_status']

# Separate the X variable, the features
X = lending_df.drop(columns=['loan_status'])

In [4]:
# Review the y variable Series

y[:5]
#display(y.head())
#display(y.tail())

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [5]:
# Review the X variable DataFrame
#X[:5]
X.head()
# display(X.head())
# display(X.tail())

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Check the balance of our target values
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [34]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
lr_model = logistic_regression_model.fit(X_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [35]:
# Make a prediction using the testing data
testing_predictions = lr_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [39]:
# Print the balanced_accuracy score of the model

print(f"Training Data Score: {lr_model.score(X_train,y_train)}")
print(f"Testing Data Score: {lr_model.score(X_test,y_test)}")
print(f"Balanced Accuracy Score: {balanced_accuracy_score(y_test, testing_predictions)}")
print(f"The balanced accuracy with the original data is {testing_accuracy* 100:.2f}%")


Training Data Score: 0.9914878250103177
Testing Data Score: 0.9924164259182832
Balanced Accuracy Score: 0.9442676901753825
The balanced accuracy with the original data is 94.43%


In [40]:
# Generate a confusion matrix for the model
testing_confusion_matrix = confusion_matrix(y_test, testing_prediction)

print(f'The confusion matrix with the original data is:\n{testing_confusion_matrix}')

The confusion matrix with the original data is:
[[18679    80]
 [   67   558]]


In [41]:
# Print the classification report for the model
testing_report = classification_report(y_test, testing_prediction)

#View the results
print(f'The classification report with the original data shows:\n{testing_report}')


The classification report with the original data shows:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      0.89      0.88       625

    accuracy                           0.99     19384
   macro avg       0.94      0.94      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 
HEALTHY LOANS: For healthy loans, the precision is 1.00, the recall is 1.00 and the f1-score is 1.00, meaning the model is performing exceptionally well in identifying healthy loans without missing any. This scenario is ideal, indicating a high level of confidence in the model's ability to identify healthy loans. 

HIGH-RISK LOANS: The precision is 0.87 for high-risk loans, so the remaining 13% are false-positives. The recall is 0.89, so the remaining 11% are false negatives. The f1-score is 0.88, which suggests a good balance between precision and recall for high-risk loans. 
Therefore, the model performs moderately well for high-risk loans, but there is still room for improvement; it is highly likely that the financial field would require 95% or higher in order to retain confidence in the model. 

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [42]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros_model = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_res, y_res = ros_model.fit_resample(X_train, y_train)

In [43]:
# Count the distinct values of the resampled labels data

#y_res.value_counts()
print(y_res.value_counts())


# Check that my numbers are the same on both sides
unique_values, counts = np.unique(y_res, return_counts=True)
print("Resampled Labels: ", unique_values)
print("Label Counts: ", counts)

0    56277
1    56277
Name: loan_status, dtype: int64
Resampled Labels:  [0 1]
Label Counts:  [56277 56277]


### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [44]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
lr_ros_model = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
lr_ros_model.fit(X_res, y_res)

# Make a prediction using the testing data
testing_ros_predictions = lr_ros_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [70]:
# Print the balanced_accuracy score of the model 
print(f"Training Data Score: {lr_ros_model.score(X_res,y_res)}")
print(f"Testing Data Score: {lr_ros_model.score(X_test,y_test)}")

ros_balanced_accuracy_score = balanced_accuracy_score(y_test, testing_ros_predictions)
print(f"Balanced Accuracy Score: {ros_balanced_accuracy_score}")
print(f"The balanced accuracy with the original data is {ros_balanced_accuracy_score* 100:.2f}%")


Training Data Score: 0.994180571103648
Testing Data Score: 0.9952022286421791
Balanced Accuracy Score: 0.9959744975744975
The balanced accuracy with the original data is 99.60%


In [79]:
# Generate a confusion matrix for the model
ros_confusion_matrix = confusion_matrix(y_test, testing_ros_predictions)
print(f'The confusion matrix with the oversampled data is:\n{ros_confusion_matrix}')

The confusion matrix with the oversampled data is:
[[18668    91]
 [    2   623]]


In [80]:
# Print the classification report for the model
ros_classification_report = classification_report(y_test, testing_ros_predictions)

#View the results
print(f'The classification report with the oversampled data shows:\n{ros_classification_report}')



The classification report with the oversampled data shows:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      1.00      0.93       625

    accuracy                           1.00     19384
   macro avg       0.94      1.00      0.96     19384
weighted avg       1.00      1.00      1.00     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model, fit with the oversampled data, better predicts the healthy and high-risk loan labels than our first report. The healthy loans precision, recall, and f1-score remain the same as before; likewise, the high-risk loan labels have the same precision measurement. However, the recall has gone up 0.01 to a perfect "1.00", while the oversampled f1-score has increase the most, by 0.05, bringing it's score to "0.93". These changes make the logistic regression model, fit with the oversampled data, does a better job in catching the incorrect labelling of high-risk loans as healthy..